Frontend/src/pages/register/components/RegistrationPanel.jsx

```diff
@@ -1,48 +1,141 @@
 import submitEventRegistration from '../../../api/registration/post/submit_registration'
 import { setMessage } from '../../../ui/events/messages'
+import LoadingMessage from '../../../ui/messages/loadingMessage'
 import styles from './panel.module.scss'
 
 export default function RegistrationPanel() {
   const { user } = useContext(AuthContext)
   const { competitionInfo } = useContext(CompetitionContext)
   const [comment, setComment] = useState('')
   const [selectedEvents, setSelectedEvents] = useState([])
-  return (
+  const [registration, setRegistration] = useState({})
+  const queryClient = useQueryClient()
+  const { data: registrationRequest, isLoading } = useQuery({
+    queryKey: ['registration', user, competitionInfo.id],
+    queryFn: () => getSingleRegistration(user, competitionInfo.id),
+  })
+  useEffect(() => {
+    if (
+      registrationRequest &&
+      registrationRequest.registration.registration_status
+    ) {
+      setRegistration(registrationRequest.registration)
+      setComment(registrationRequest.registration.comment)
+      setSelectedEvents(registrationRequest.registration.event_ids)
+    }
+  }, [registrationRequest])
+  const { mutate: updateRegistrationMutation } = useMutation({
+    mutationFn: updateRegistration,
+    onError: (data) => {
+      setMessage(
+        'Registration update failed with error: ' + data.error,
+        'negative'
+      )
+    },
+    onSuccess: (data) => {
+      setMessage('Registration update succeeded', 'positive')
+      queryClient.setQueryData(['registration', competitionInfo.id, user], data)
+    },
+  })
+  const { mutate: createRegistrationMutation } = useMutation({
+    mutationFn: submitEventRegistration,
+    onError: (data) => {
+      setMessage('Registration failed with error: ' + data.error, 'negative')
+    },
+    onSuccess: (_) => {
+      // We can't update the registration yet, because there might be more steps needed
+      // And the Registration might still be processing
+      setMessage('Registration submitted successfully', 'positive')
+    },
+  })
+
+  return isLoading ? (
+    <LoadingMessage />
+  ) : (
     <div className={styles.panel}>
-      <EventSelector
-        handleEventSelection={setSelectedEvents}
-        events={competitionInfo.event_ids}
-        initialSelected={[]}
-        size="2x"
+      {registration.registration_status ? (
+        <>
+          <h3>You have registered for {competitionInfo.name}</h3>
+          {/* Really weird bug here: Only if I have the console.log statement here (or any other statement referencing it), initialSelected, will work correctly?? */}
+          {/* eslint-disable-next-line no-console */}
+          {console.log(registration.event_ids)}
+          <EventSelector
+            handleEventSelection={setSelectedEvents}
+            events={competitionInfo.event_ids}
+            initialSelected={registration.event_ids}
```

Comment:
I don't understand this, Is there meant to be a comma after 'initialSelected'?

But I think I know what's going on. This child component won't render while the query is loading (due to the `isLoading? <LoadingMessage> : ...`). But, once the query is done loading, this child will render. However, because an effect is being used to call `setRegistration`, and effects don't happen until the next render cycle, `registration.event_ids` only has the 'correct' value on the second render of `EventSelector`. And of course `EventSelector` only uses the `intialSelected` value on first render, then ignores it subsequently.

This is one of the reasons `useEffect` is generally not good, and why it should be avoided unless using queries to populate form inputs (as discussed on slack). If we were directly referencing the `data` of the query, it would be available on the first render.

The quickest fix would be to change `const [registration, setRegistration] = useState(null) // null initial state` and `return (isLoading || !registration) ? ( ...`, but that feels unclean because the whole point of `useQuery`'s `isLoading` is to avoid needing to explicitly check data is present.

A (probably) slightly better fix is it switch `EventSelector` to a controlled input. So instead of passing `initialSelected` and it having an internal state, you pass in `selected` and `handEventSelection` and it doesn't need any internal state. You can also make it more general so that it can be used as a controlled or uncontrolled component, I can show an example if you want.

No idea why the console log would 'fix' this bug :thinking:

Reply:
Ok, then I will change this in the next wca-component iteration